pkg/controller/pipelinerun/controller_test.go

```diff
@@ -0,0 +1,175 @@
+	}
+
+	ps := []*v1alpha1.Pipeline{{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-pipeline",
+			Namespace: "foo",
+		}},
+	}
+
+	tcs := []struct {
+		name           string
+		pipelineRun    string
+		runtimeSuccess bool
+		expectedRecord string
+	}{
+		{"success", "foo/test-pipeline-run-success", true, "Normal Synced Found pipeline test-pipeline"},
+		{"fail", "test-pipeline-run-fail", false, ""},
+	}
+	ch := make(chan string, 10)
+	defer close(ch)
+	setUpMocks(prs, ps, ch)
+
+	c := NewController(client, pipelineClient, logging.GetContextLogger(t.Name()).Logger).(*Controller)
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			old := runtime.ErrorHandlers
+			var actualErr error
+			defer func() { runtime.ErrorHandlers = old }()
+			runtime.ErrorHandlers = []func(error){
+				func(err error) {
+					actualErr = err
+				},
+			}
+			c.workqueue.Add(tc.pipelineRun)
+			if !c.processNextWorkItem() {
+				t.Errorf("expected to process work item successfuly")
+			}
+			if tc.runtimeSuccess != (actualErr == nil) {
+				t.Errorf("expected to see error %t. Got error %v", tc.runtimeSuccess, actualErr)
+			}
+			if actualErr == nil {
+				// Check message is recorded.
+				msg := <-ch
+				if msg != tc.expectedRecord {
+					t.Errorf("expected to see %s event got %s", tc.expectedRecord, msg)
+				}
+			}
+		})
+	}
+	c.workqueue.ShutDown()
+}
+
+// sets up all the mocks for testing
+func setUpMocks(prs []*v1alpha1.PipelineRun, ps []*v1alpha1.Pipeline, ch chan string) {
+	runsListerFunc = func(i pi.PipelineRunInformer) listers.PipelineRunLister {
+		return &mockPipelineRunsLister{runs: prs}
+	}
+	pipelinesListerFunc = func(i pi.PipelineInformer) listers.PipelineLister {
+		return &mockPipelinesLister{p: ps}
+	}
+	getRecordFunc = func(k kubernetes.Interface, l *zap.SugaredLogger) record.EventRecorder {
+		return &record.FakeRecorder{
+			Events: ch,
+		}
+	}
+}
+
+type fakeStruct struct {
+}
+
+type mockPipelineRunsLister struct {
+	runs []*v1alpha1.PipelineRun
+}
+
+func (m *mockPipelineRunsLister) List(selector labels.Selector) (ret []*v1alpha1.PipelineRun, err error) {
+	return m.runs, nil
+}
+
+func (m *mockPipelineRunsLister) PipelineRuns(namespace string) listers.PipelineRunNamespaceLister {
+	return m
+}
+
+func (m *mockPipelineRunsLister) Get(name string) (*v1alpha1.PipelineRun, error) {
+	for _, run := range m.runs {
+		if run.Name == name {
+			return run, nil
+		}
+	}
+	return nil, fmt.Errorf("%s not found", name)
+}
+
+type mockPipelinesLister struct {
+	p []*v1alpha1.Pipeline
+}
+
+func (m *mockPipelinesLister) List(selector labels.Selector) (ret []*v1alpha1.Pipeline, err error) {
+	return m.p, nil
+}
+
+func (m *mockPipelinesLister) Pipelines(namespace string) listers.PipelineNamespaceLister {
+	return m
+}
+
+func (m *mockPipelinesLister) Get(name string) (*v1alpha1.Pipeline, error) {
+	for _, p := range m.p {
+		if p.Name == name {
+			return p, nil
+		}
+	}
+	return nil, fmt.Errorf("%s not found", name)
+}
```

Comment:
how many of these functions are reusable for other tests, and as folks add more tests, which of these will they need to duplicated?

could be really handy to have some docstrings or markdown to add some guidance for folks writing more unit tests

Reply:
I usually follow this style
1. Keep test mocks closer to the test mocks,  so you do need to look in to another file.
Sacrifice repeatability over readability.
2. In case some other tests need the same func, then and only then move what is required out in to testutil.

So, to answer your questions, the next time we  re-use them we should decide which ones to move out in the util.  
I can definitely add some docs to some of these are private methods or add a Readme in the Developer guide instead

Reply:
sounds good to me! my remaining ask would be some guidance for future test authors on how to use this as a starting point for a new test - even if that means literally copying this file and changing things (which it sounds like is the case at the moment, and im totally fine with that)

> only then move what is required out in to testutil.

as long as we don't call it `testutil` ;) https://blog.golang.org/package-names but that is probably a discussion for another day!

Reply:
Will add these in the coming update.